### PR TITLE
Implement ldns-update and dnst update

### PR DESCRIPTION
**Pull request by tertsdiepraam**
I guess documentation is a bit lacking and @wtoorop has updated the docs for ldns, which I should still incorporate. But the actual code is ready for review.

### CODE REVIEW

File: src/commands/update.rs [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+                        let key = Key::new(
+                            tsig.algorithm,
+                            &tsig.key,
+                            tsig.name.clone(),
+                            None,
+                            None,
+                        )
+                        .unwrap();
+                        let msg =
+                            RequestMessageMulti::new(msg.clone()).unwrap();
```
Comment (ximon18) [inline src/commands/update.rs:418]: What's the reason for using Multi here? I would only expect to see that if XFR responses are expected.

Comment (tertsdiepraam) [reply src/commands/update.rs:418]: It was the only thing that worked with the types...

Comment (tertsdiepraam) [reply src/commands/update.rs:418]: That'd be great! I think there's a missing impl somewhere but forgot to report that back.

Comment (ximon18) [reply src/commands/update.rs]: Done.

Comment (tertsdiepraam) [reply src/commands/update.rs]: I'm confused, there's still `RequestMessageMulti` now right?

File: src/commands/update.rs [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+
+    async fn send_update_with_tsig(
+        &self,
+        msg: RequestMessage<Vec<u8>>,
+        key: Key,
+        socket: SocketAddr,
+    ) -> Result<Message<Bytes>, domain::net::client::request::Error> {
+        let tcp_conn = TcpStream::connect(socket).await.unwrap();
+        let (connection, transport) =
+            stream::Connection::<_, RequestMessageMulti<Vec<u8>>>::new(tcp_conn);
```
Comment (tertsdiepraam) [inline src/commands/update.rs:454]: Here

**Comment by tertsdiepraam** [comment]
This could probably use more tests, but it seems to mostly work?

File: src/commands/update.rs [hunk 3]
```diff
@@ -0,0 +1,676 @@
...
+            .await?;
+
+        let mut nsnames = response
+            .answer()?
+            .limit_to_in::<Ns<_>>()
+            .map(|ns| Ok(ns?.data().nsdname().to_name::<Vec<u8>>()))
+            .collect::<Result<Vec<_>, Error>>()?;
+
+        // The mname should be tried first according to RFC2136 4.3
+        // so we put that nsname first in the list.
```
Comment (mozzieongit) [inline src/commands/update.rs:333]:
```suggestion
        // The MNAME should be tried first according to RFC2136 4.3
        // so we put that NSNAME first in the list.
```

File: src/commands/update.rs [hunk 4]
```diff
@@ -0,0 +1,676 @@
...
+        };
+
+        let mut key;
+        let mut algorithm;
+        if let Some((k, a)) = rest.split_once(':') {
+            key = k;
+            algorithm = a;
+        } else {
+            key = rest;
+            algorithm = "hmac-sha512";
```
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/commands/update.rs:74]: This ldns tool doesn't have a default. Which makes me wonder if this should even have a default.

File: src/commands/update.rs [hunk 5]
```diff
@@ -0,0 +1,676 @@
...
+        let Some(soa) = sections.next() else {
+            return Err("no SOA found".into());
+        };
+
+        let soa_mname: Name<Vec<u8>> = soa?.data().mname().to_name();
+
+        // Step 2 - find SOA MNAME IP address, add to resolver
+        let response = resolver
+            .query(Question::new(&soa_mname, Rtype::A, Class::IN))
+            .await?;
```
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/commands/update.rs]: Good catch!

File: src/commands/update.rs [hunk 6]
```diff
@@ -0,0 +1,676 @@
...
+    ip: OptionIpAddr,
+
+    /// Zone to update
+    #[arg(long = "zone")]
+    zone: Option<Name<Vec<u8>>>,
+
+    /// TSIG credentials for the UPDATE packet
+    #[arg(long = "tsig", value_name = "name:key[:algo]")]
+    tsig: Option<TSigInfo>,
+}
```
[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/commands/update.rs:47]: @mozzieongit and I decided to defer this and make an issue out of this.

File: src/commands/update.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (mozzieongit) [reply src/commands/update.rs]: Good point, maybe you're right. 
I think in the end it doesn't matter much. For some people the default will save a few characters of typing, for others it doesn't do much, other that TSIG might fail if they forget to provide the correct algorithm. Although that could already be considered annoying.

File: src/commands/update.rs [hunk 7]
```diff
@@ -0,0 +1,676 @@
...
+        let resolver = env.stub_resolver_from_conf(conf).await;
+
+        let response = resolver
+            .query(Question::new(&name, Rtype::SOA, Class::IN))
+            .await?;
+
+        let mut authority = response.authority()?.limit_to::<Soa<_>>();
+        let Some(soa) = authority.next() else {
+            return Err("no SOA record found hello".into());
+        };
```
[the remark this answers is not in the document]
Comment (mozzieongit) [reply src/commands/update.rs]: Ah I get it now. It's to make sure we have the correct SOA record from the primary name server (and not some potentially outdated one from a secondary).

However, I believe to see an issue here (only read the code of both ldns and dnst, not tested): If `name` is the zone apex, the SOA record won't be in the authority section, but the answer section. So this is probably a bug in the original ldns-update then.

[the remark this answers is not in the document]
Comment (tertsdiepraam) [reply src/commands/update.rs]: Ahh that makes sense. I suppose I'll update this to look at both then?

**Review by bal-e — APPROVED** [review]
Nothing's leaping out at me.  It looks pretty good!

File: src/parse.rs [hunk 8]
```diff
@@ -0,0 +1,66 @@
+use std::str::FromStr;
+
+use domain::{
+    tsig::{Algorithm, KeyName},
+    utils::base64,
+};
```
Comment (mozzieongit) [inline src/parse.rs:6]:
```suggestion
use domain::tsig::{Algorithm, KeyName};
use domain::utils::base64;
```

**Review by mozzieongit — APPROVED** [review]
Looks good :+1: 
Only a tiny nit

**Pull request merged**